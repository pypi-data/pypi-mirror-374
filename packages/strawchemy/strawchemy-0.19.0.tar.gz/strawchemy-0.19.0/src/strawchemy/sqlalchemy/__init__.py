"""SQLAlchemy integration for Strawchemy.

This package provides SQLAlchemy-based repository implementations for use with
Strawberry GraphQL.
"""

from __future__ import annotations

from .repository import SQLAlchemyGraphQLAsyncRepository, SQLAlchemyGraphQLRepository, SQLAlchemyGraphQLSyncRepository

__all__ = ("SQLAlchemyGraphQLAsyncRepository", "SQLAlchemyGraphQLRepository", "SQLAlchemyGraphQLSyncRepository")
