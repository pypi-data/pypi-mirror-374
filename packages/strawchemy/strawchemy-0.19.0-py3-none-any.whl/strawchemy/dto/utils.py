"""Utilities for DTO configuration and type checking.

This module provides utility functions for configuring Data Transfer Objects (DTOs)
and performing type checks, specifically for optional type hints.

It exports the following:

- `config`: Configures a DTOConfig object for a specific purpose.
- `field`: Configures a DTOFieldConfig object for a specific purpose.
- `is_type_hint_optional`: Checks if a type hint is optional.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from .constants import DTO_INFO_KEY
from .types import DTOConfig, DTOFieldConfig, DTOScope, ExcludeFields, IncludeFields, Purpose, PurposeConfig

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping

__all__ = (
    "PRIVATE",
    "READ_ONLY",
    "WRITE_ONLY",
    "config",
    "field",
    "read_all_config",
    "read_all_partial_config",
    "read_partial",
    "write_all_config",
    "write_all_partial_config",
)


def config(
    purpose: Purpose,
    include: Optional[IncludeFields] = None,
    exclude: Optional[ExcludeFields] = None,
    partial: Optional[bool] = None,
    type_map: Optional[Mapping[Any, Any]] = None,
    aliases: Optional[Mapping[str, str]] = None,
    alias_generator: Optional[Callable[[str], str]] = None,
    scope: Optional[DTOScope] = None,
    tags: Optional[set[str]] = None,
) -> DTOConfig:
    config = DTOConfig(purpose, alias_generator=alias_generator, scope=scope)
    if exclude:
        config.exclude = exclude
    if include:
        config.include = include
    if type_map:
        config.type_overrides = type_map
    if aliases:
        config.aliases = aliases
    if partial is not None:
        config.partial = partial
    if tags:
        config.tags = tags
    return config


def field(
    purposes: Optional[set[Purpose]] = None,
    default_config: Optional[PurposeConfig] = None,
    configs: Optional[dict[Purpose, PurposeConfig]] = None,
) -> dict[str, DTOFieldConfig]:
    return {
        DTO_INFO_KEY: DTOFieldConfig(
            purposes=purposes if purposes is not None else {Purpose.READ, Purpose.WRITE},
            default_config=default_config or PurposeConfig(),
            configs=configs or {},
        ),
    }


read_partial = DTOConfig(Purpose.READ, partial=True)
read_all_config = DTOConfig(Purpose.READ, include="all")
read_all_partial_config = DTOConfig(Purpose.READ, include="all", partial=True)
write_all_config = DTOConfig(Purpose.WRITE, include="all")
write_all_partial_config = DTOConfig(Purpose.WRITE, include="all", partial=True)

READ_ONLY = field({Purpose.READ})
WRITE_ONLY = field({Purpose.WRITE})
PRIVATE = field(set())
