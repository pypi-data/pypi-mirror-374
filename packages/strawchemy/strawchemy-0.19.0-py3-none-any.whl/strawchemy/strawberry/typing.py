from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections.abc import Callable

    from sqlalchemy import Select
    from strawberry import Info
    from strawberry.types.base import WithStrawberryObjectDefinition
    from strawchemy.graph import Node
    from strawchemy.sqlalchemy.typing import AnyAsyncSession, AnySyncSession
    from strawchemy.strawberry.dto import GraphQLFieldDefinition, QueryNodeMetadata, StrawchemyDTOAttributes
    from strawchemy.validation.pydantic import MappedPydanticGraphQLDTO

    from .dto import (
        AggregateDTO,
        FilterFunctionInfo,
        GraphQLFilterDTO,
        MappedStrawberryGraphQLDTO,
        OrderByDTO,
        OutputFunctionInfo,
        UnmappedStrawberryGraphQLDTO,
    )

__all__ = (
    "AnySessionGetter",
    "AsyncSessionGetter",
    "FilterStatementCallable",
    "StrawchemyTypeWithStrawberryObjectDefinition",
    "SyncSessionGetter",
)


_T = TypeVar("_T")
QueryObject = TypeVar("QueryObject", bound="Any")
GraphQLFilterDTOT = TypeVar("GraphQLFilterDTOT", bound="GraphQLFilterDTO")
AggregateDTOT = TypeVar("AggregateDTOT", bound="AggregateDTO")
GraphQLDTOT = TypeVar("GraphQLDTOT", bound="GraphQLDTO[Any]")
OrderByDTOT = TypeVar("OrderByDTOT", bound="OrderByDTO")

AggregationFunction = Literal["min", "max", "sum", "avg", "count", "stddev_samp", "stddev_pop", "var_samp", "var_pop"]
AggregationType = Literal[
    "sum", "numeric", "min_max_datetime", "min_max_date", "min_max_time", "min_max_string", "min_max_numeric"
]

GraphQLType = Literal["input", "object", "interface", "enum"]
AsyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnyAsyncSession]"
SyncSessionGetter: TypeAlias = "Callable[[Info[Any, Any]], AnySyncSession]"
AnySessionGetter: TypeAlias = "AsyncSessionGetter | SyncSessionGetter"
FilterStatementCallable: TypeAlias = "Callable[[Info[Any, Any]], Select[tuple[Any]]]"
GraphQLPurpose: TypeAlias = Literal[
    "type",
    "aggregate_type",
    "create_input",
    "update_by_pk_input",
    "update_by_filter_input",
    "filter",
    "aggregate_filter",
    "order_by",
    "upsert_update_fields",
    "upsert_conflict_fields",
]
FunctionInfo: TypeAlias = "FilterFunctionInfo | OutputFunctionInfo"
StrawberryGraphQLDTO: TypeAlias = "MappedStrawberryGraphQLDTO[_T] | UnmappedStrawberryGraphQLDTO[_T]"
GraphQLDTO: TypeAlias = "StrawberryGraphQLDTO[_T] | MappedPydanticGraphQLDTO[_T]"
MappedGraphQLDTO: TypeAlias = "MappedStrawberryGraphQLDTO[_T] | MappedPydanticGraphQLDTO[_T]"
AnyMappedDTO: TypeAlias = "MappedStrawberryGraphQLDTO[Any] | MappedPydanticGraphQLDTO[Any]"
QueryNodeType: TypeAlias = "Node[GraphQLFieldDefinition, QueryNodeMetadata]"

if TYPE_CHECKING:

    class StrawchemyTypeWithStrawberryObjectDefinition(StrawchemyDTOAttributes, WithStrawberryObjectDefinition): ...
