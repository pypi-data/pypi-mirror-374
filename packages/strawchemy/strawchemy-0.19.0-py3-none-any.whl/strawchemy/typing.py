from __future__ import annotations

import sys
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Protocol, Union

from typing_extensions import TypeAlias

if sys.version_info >= (3, 10):
    from types import UnionType

    UNION_TYPES = (Union, UnionType)
else:
    UNION_TYPES = (Union,)


if TYPE_CHECKING:
    from . import StrawchemyAsyncRepository, StrawchemySyncRepository

__all__ = ("UNION_TYPES", "AnyRepository", "DataclassProtocol", "SupportedDialect")


class DataclassProtocol(Protocol):
    __dataclass_fields__: ClassVar[dict[str, Any]]


AnyRepository: TypeAlias = "type[StrawchemySyncRepository[Any] | StrawchemyAsyncRepository[Any]]"
SupportedDialect: TypeAlias = Literal["postgresql", "mysql", "sqlite"]
"""Must match SQLAlchemy dialect."""
