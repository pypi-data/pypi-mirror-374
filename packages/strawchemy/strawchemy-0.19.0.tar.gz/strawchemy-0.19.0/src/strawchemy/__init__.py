"""Custom DTO implementation."""

from __future__ import annotations

from .config.base import StrawchemyConfig
from .mapper import Strawchemy
from .sqlalchemy.hook import QueryHook
from .strawberry import ModelInstance
from .strawberry.mutation.input import Input
from .strawberry.mutation.types import (
    ErrorType,
    RequiredToManyUpdateInput,
    RequiredToOneInput,
    ToManyCreateInput,
    ToManyUpdateInput,
    ToOneInput,
    ValidationErrorType,
)
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository
from .validation.base import InputValidationError

__all__ = (
    "ErrorType",
    "Input",
    "InputValidationError",
    "ModelInstance",
    "QueryHook",
    "RequiredToManyUpdateInput",
    "RequiredToOneInput",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemyConfig",
    "StrawchemySyncRepository",
    "ToManyCreateInput",
    "ToManyUpdateInput",
    "ToOneInput",
    "ValidationErrorType",
)
