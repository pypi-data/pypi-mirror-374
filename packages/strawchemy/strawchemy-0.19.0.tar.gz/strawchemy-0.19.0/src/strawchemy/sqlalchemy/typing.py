from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections import OrderedDict
    from collections.abc import Callable

    from sqlalchemy import Function
    from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session
    from sqlalchemy.orm import DeclarativeBase, Session, scoped_session
    from sqlalchemy.sql import SQLColumnExpression
    from strawchemy.strawberry.dto import OrderByEnum
    from strawchemy.strawberry.filters.base import GraphQLComparison

    from ._executor import QueryExecutor
    from .hook import QueryHook


__all__ = (
    "AnyAsyncSession",
    "AnySession",
    "AnySyncSession",
    "DeclarativeSubT",
    "DeclarativeT",
    "FunctionGenerator",
    "QueryExecutorT",
    "QueryHookCallable",
    "RelationshipSide",
    "SessionT",
    "StatementType",
)

DeclarativeT = TypeVar("DeclarativeT", bound="DeclarativeBase")
DeclarativeSubT = TypeVar("DeclarativeSubT", bound="DeclarativeBase")
QueryHookDeclarativeT = TypeVar("QueryHookDeclarativeT", bound="DeclarativeBase")
SessionT = TypeVar("SessionT", bound="AnySession")
QueryExecutorT = TypeVar("QueryExecutorT", bound="QueryExecutor[Any]")

RelationshipSide: TypeAlias = Literal["parent", "target"]
StatementType = Literal["lambda", "select"]
FunctionGenerator: TypeAlias = "Callable[..., Function[Any]]"
QueryHookCallable: TypeAlias = "QueryHook[QueryHookDeclarativeT]"
FilterMap: TypeAlias = "OrderedDict[tuple[type[Any], ...], type[GraphQLComparison]]"
AnySyncSession: TypeAlias = "Session | scoped_session[Session]"
AnyAsyncSession: TypeAlias = "AsyncSession | async_scoped_session[AsyncSession]"
AnySession: TypeAlias = "AnySyncSession | AnyAsyncSession"
OrderBySpec: TypeAlias = "tuple[SQLColumnExpression[Any], OrderByEnum]"
