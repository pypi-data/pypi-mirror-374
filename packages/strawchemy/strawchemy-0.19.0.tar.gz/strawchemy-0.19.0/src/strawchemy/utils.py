from __future__ import annotations

import inspect
import re
import sys
from typing import TYPE_CHECKING, Any, Optional, Union, get_args, get_origin

from strawchemy.typing import UNION_TYPES

if TYPE_CHECKING:
    from re import Pattern

__all__ = (
    "camel_to_snake",
    "is_type_hint_optional",
    "non_optional_type_hint",
    "snake_keys",
    "snake_to_camel",
    "snake_to_lower_camel_case",
)

_camel_to_snake_pattern: Pattern[str] = re.compile(r"((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))")


def camel_to_snake(string: str) -> str:
    """Convert a camelcased string to snake case.

    See: https://stackoverflow.com/a/12867228
    """
    return _camel_to_snake_pattern.sub(r"_\1", string).lower()


def snake_to_camel(string: str) -> str:
    """Convert string to camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    return "".join(x.capitalize() for x in string.lower().split("_"))


def snake_to_lower_camel_case(snake_str: str) -> Any:
    """Convert string to lower camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    camel_string: str = snake_to_camel(snake_str)
    return snake_str[0].lower() + camel_string[1:]


def snake_keys(value: dict[str, Any]) -> dict[str, Any]:
    """Recursively convert dict keys to from camel case to snake case."""
    res: dict[Any, Any] = {}
    for k, v in value.items():
        to_snake: str = camel_to_snake(k)
        if isinstance(v, (list, tuple)):
            res[to_snake] = [snake_keys(el) for el in v]
        elif isinstance(v, dict):
            res[to_snake] = snake_keys(v)
        else:
            res[to_snake] = v
    return res


def non_optional_type_hint(type_hint: Any) -> Any:
    origin, args = get_origin(type_hint), get_args(type_hint)
    if origin is Optional:
        return args
    if origin in UNION_TYPES:
        union_args = tuple([arg for arg in args if arg not in (None, type(None))])
        if len(union_args) == 1:
            return union_args[0]
        return Union[union_args]
    return type_hint


def is_type_hint_optional(type_hint: Any) -> bool:
    """Whether the given type hint is considered as optional or not.

    Returns:
        `True` if arguments of the given type hint are optional

    Three cases are considered:
    ```
        Optional[str]
        Union[str, None]
        str | None
    ```
    In any other form, the type hint will not be considered as optional
    """
    origin = get_origin(type_hint)
    if origin is None:
        return False
    if origin is Optional:
        return True
    if origin in UNION_TYPES:
        args = get_args(type_hint)
        return any(arg is type(None) for arg in args)
    return False


def get_annotations(obj: Any) -> dict[str, Any]:
    """Get the annotations of the given object."""
    if sys.version_info >= (3, 10):
        return inspect.get_annotations(obj)
    return obj.__dict__.get("__annotations__", {}) if isinstance(obj, type) else getattr(obj, "__annotations__", {})
