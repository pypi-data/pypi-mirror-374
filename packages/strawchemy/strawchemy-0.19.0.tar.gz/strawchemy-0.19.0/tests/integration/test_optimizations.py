from __future__ import annotations

import re
import warnings
from typing import TYPE_CHECKING

import pytest

from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .fixtures import QueryTracker

if TYPE_CHECKING:
    from strawchemy.typing import SupportedDialect

    from syrupy.assertion import SnapshotAssertion

pytestmark = [pytest.mark.integration]


@pytest.mark.parametrize(
    "query",
    [
        pytest.param(
            """
        {
            colors(orderBy: { fruitsAggregate: { count: ASC } }) {
                fruitsAggregate {
                    count
                }
            }
        }
        """,
            id="output-order-by",
        ),
        pytest.param(
            """
        {
            colors(filter: { fruitsAggregate: { count: { predicate: { gt: 0 } } } }) {
                fruitsAggregate {
                    count
                }
            }
        }
        """,
            id="output-filter",
        ),
        pytest.param(
            """
        {
            colors(
                filter: { fruitsAggregate: { count: { predicate: { gt: 0 } } } },
                orderBy: { fruitsAggregate: { avg: { waterPercent: ASC } } }
            ) {
                fruits {
                    id
                }
            }
        }
        """,
            id="filter-order-by",
        ),
        pytest.param(
            """
        {
            colors(
                filter: { fruitsAggregate: { avg: { arguments: [waterPercent] predicate: { gt: 0 } } } },
                orderBy: { fruitsAggregate: { avg: { waterPercent: ASC } } }
            ) {
                fruits {
                    id
                }
            }
        }
        """,
            id="filter-order-by-same-aggregation",
        ),
        pytest.param(
            """
        {
            colors {
                fruitsAggregate {
                    max {
                        sweetness
                        waterPercent
                        name
                    }
                }
            }
        }
        """,
            id="output-multiple-aggregations",
        ),
        pytest.param(
            """
        {
            colors(
                filter: {
                    fruitsAggregate: {
                        sum: { arguments: [sweetness], predicate: { gt: 0 } },
                        avg: { arguments: [waterPercent], predicate: { gt: 0 } }
                    }
                }
            ) {
                fruits {
                    id
                }
            }
        }
        """,
            id="filter-multiple-aggregations",
        ),
        pytest.param(
            """
        {
            colors(
                orderBy: { fruitsAggregate: { sum: { sweetness: ASC }, avg: { waterPercent: ASC } } }
            ) {
                fruits {
                    id
                }
            }
        }
        """,
            id="order-by-multiple-aggregations",
        ),
    ],
)
async def test_aggregation_computation_is_reused(
    query: str, any_query: AnyQueryExecutor, query_tracker: QueryTracker, dialect: SupportedDialect
) -> None:
    """Test that aggregation computation is reused when filtering and ordering by the same aggregation."""
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        result = await maybe_async(any_query(query))

    assert not result.errors
    assert result.data

    assert query_tracker.query_count == 1
    if dialect == "postgresql":
        assert query_tracker[0].statement_str.count("JOIN LATERAL (") == 1
    else:
        assert query_tracker[0].statement_str.count("aggregation_0 AS ") == 1
        assert not re.match(r" aggregation_[1-9]\d* AS $", query_tracker[0].statement_str)


@pytest.mark.parametrize(
    ("query", "inner_join_expected"),
    [
        pytest.param(
            """
                    {
                        colors(filter: { fruits: { sweetness: { gt: 1 } } }) {
                            fruits {
                                sweetness
                            }
                        }
                    }
                """,
            True,
            id="inner-join-rewrite",
        ),
        pytest.param(
            """
                    {
                        colors(filter: { createdAt: { gt: "1220-01-01T00:00:00" } }) {
                            fruits {
                                sweetness
                            }
                        }
                    }
                """,
            False,
            id="no-inner-join-rewrite",
        ),
    ],
)
@pytest.mark.snapshot
async def test_inner_join_rewriting(
    query: str,
    inner_join_expected: bool,
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    """Test that if WHERE condition only references columns from the null-supplying side of the join, use an inner join."""
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot

    if inner_join_expected:
        assert "LEFT OUTER JOIN" not in query_tracker[0].statement_str
        assert query_tracker[0].statement_str.count("JOIN") == 1
    else:
        assert query_tracker[0].statement_str.count("LEFT OUTER JOIN") == 1
