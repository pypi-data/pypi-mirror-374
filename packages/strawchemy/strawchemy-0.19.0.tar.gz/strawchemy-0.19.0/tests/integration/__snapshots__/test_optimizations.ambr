# serializer version: 1
# name: test_inner_join_rewriting[session-tracked-async-inner-join-rewrite-aiosqlite_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > ?
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-inner-join-rewrite-asyncmy_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
   INNER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > %s
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-inner-join-rewrite-asyncpg_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > $1::INTEGER
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-inner-join-rewrite-psycopg_async_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > %(sweetness_1)s::INTEGER
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-no-inner-join-rewrite-aiosqlite_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > ?
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-no-inner-join-rewrite-asyncmy_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > %s
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-no-inner-join-rewrite-asyncpg_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > $1::TIMESTAMP WITHOUT TIME
    ZONE
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-async-no-inner-join-rewrite-psycopg_async_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > %(created_at_1)s::TIMESTAMP WITHOUT TIME
    ZONE
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-sync-inner-join-rewrite-psycopg_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > %(sweetness_1)s::INTEGER
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-sync-inner-join-rewrite-sqlite_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color__fruits.sweetness > ?
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-sync-no-inner-join-rewrite-psycopg_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > %(created_at_1)s::TIMESTAMP WITHOUT TIME
    ZONE
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
# name: test_inner_join_rewriting[session-tracked-sync-no-inner-join-rewrite-sqlite_engine]
  '''
  SELECT color__fruits.sweetness,
         color__fruits.id,
         color.id AS id_1
    FROM color AS color
    LEFT OUTER JOIN fruit AS color__fruits
      ON color.id = color__fruits.color_id
   WHERE color.created_at > ?
   ORDER BY color.id ASC,
            color__fruits.id ASC
  '''
# ---
